---
title: "Creating a summarized Experiment & perform a differential gene expression analysis"
author: "Marc A.T. Teunis"
date: "`r Sys.Date()`"
output: html_document
---

```{r, root, include=FALSE}
## defines the root of the project for later use
require("rprojroot") || utils::install.packages("rprojroot")
library(rprojroot)
root <- find_root_file(criterion = is_rstudio_project)
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE
                      , 
                       warning = FALSE, 
                       error = FALSE, 
                       message = FALSE
                      ,
                       results = 'hide'
                      # fig.show = 'hide'
                      )
```

# Packages
```{r}
library(tidyverse)
#source("http://bioconductor.org/workflows.R")
#workflowInstall("RNAseq123")
library(RNAseq123)
library(lattice)
library(Biobase)
#install.packages("pheatmap")
library(pheatmap)
##  install.packages("xlsx")
#library(xlsx)
library(gdata)
library(SummarizedExperiment)

```

## Reading data 

Data was obtained from Dr. Jean Paul Ten Klooster (date: 06-02-2017)
This file contains the sample information (columns or also called colData), the featureData (or `gene_id` names -> `ensembl_id`)

The files that belong to this dataset are called:
"./data/jp_counts.txt" (the count data)
"./data/jp_coldata.txt (the sample information)
"./data/jp_features.txt (gene_ensembl ids and corresponding gene names (symbol))
"./data/jp_miame.txt" (additional info - experimentData)

In this exercise you will practice to construct a summarized experiment and from individual datafiles, and from there a DESeq2 expression set, which needs to be analyzed for differential gene (mRNA) expression. 

# Prequisites: 

 - Go over the vignette and exercise for `{SummarizedExperiment}` lesson 5-2.
 - Go over the demo/theory on RNAseq analysis,lesson 6-2
 To recap:
 
# SummarizedExperiment

*SummarizedExperiment* diagram of the three
component pieces:

```{r sumexp, echo = FALSE, out.width="80%", fig.cap="The component parts of a *SummarizedExperiment* object. The 'assay' (pink block) contains the matrix of counts, the 'rowRanges' or 'rowData' (blue block) contains information about the genomic ranges and the 'colData' (green block) contains information about the samples. The highlighted line in each block represents the first row (note that the first row of 'colData' lines up with the first column of the 'assay')."}
par(mar = c(0,0,0,0))
plot(1,1,xlim = c(0,100),ylim = c(0,100),bty = "n",
     type="n",xlab="",ylab="",xaxt="n",yaxt="n")
polygon(c(45,90,90,45),c(5,5,70,70),col="pink",border=NA)
polygon(c(45,90,90,45),c(68,68,70,70),col="pink3",border=NA)
text(67.5,40,"assay")
text(67.5,35,'e.g. "counts"')
polygon(c(10,40,40,10),c(5,5,70,70),col="skyblue",border=NA)
polygon(c(10,40,40,10),c(68,68,70,70),col="skyblue3",border=NA)
text(25,40,"rowRanges,")
text(25,35,"or rowData")
polygon(c(45,90,90,45),c(75,75,95,95),col="palegreen",border=NA)
polygon(c(45,47,47,45),c(75,75,95,95),col="palegreen3",border=NA)
text(67.5,85,"colData")
```

The *SummarizedExperiment* container is diagrammed in the Figure above
and discussed in the latest Bioconductor paper [@Huber2015Orchestrating].
In our case we have created a single matrix named `"counts"` that
contains the estimated counts for each gene and sample, which is
stored in `assay`. It is also possible to store multiple matrices,
accessed with `assays`. The `rowData` for our object will keep track
of the intermediate calculations for each gene: e.g. the mean of
normalized counts, the dispersion estimates, the estimated
coefficients and their standard error, etc. We can also include
information about the location of each gene, in which case, we would
use `rowRanges` to access the *GenomicRanges* or *GenomicRangesList*
associated with each gene.The component parts of the
*SummarizedExperiment* are accessed with an R function of the same
name: `assay` (or `assays`), `rowRanges`/`rowData` and `colData`. 

## DESeq2 and edgeR
For running *DESeq2* or *edgeR* models, you can use R's formula notation to
express any fixed-effects experimental design.
Note that *DESeq2* and *edgeR* use the same formula notation as, for instance, the *lm*
function of base R. If the research aim is to determine for which
genes the effect of treatment is different across groups, then
interaction terms can be included and tested using a design such as
`~ group + treatment + group:treatment`. See the manual page for
`?results` for more examples.

# Question 1
A) Load these 4 data files (see below jp_.....txt) into R. Be sure to read the data with the appropriate function. 


```{r}
## original data

library(readr)
counts <- read_tsv(file = file.path(root, "data", "jp_counts.txt"))
pheno_data <- read_delim(file = file.path(root, "data", "jp_coldata.txt"), delim = ";")
feature_data <- read_tsv(file = file.path(root, "data", "jp_features.txt"))
miame <- read_csv2(file = file.path(root, "data", "jp_miame.txt"), col_names = TRUE)

```

B) Inspect each file when read into R (with a code chunk below). 
What can you tell about the underlying relationship between these 4 files
```{r}
files <- list(counts, pheno_data, feature_data)
lapply(files, head, 2)
miame
```

# Question 2 ExpressionSet

In this question you will need to create a SummarizedExperiment from the four individual data objects loaded above.

A) Create a summarized experiment formal class object

**TIPS**

 - See the Vignette for the package `{SummarizedExperiment}`
 - We are `contructing` a new object here...
 - make sure that the colnames of the counts DataFrame are equal to the rownames of the colData, correct if neccessary
 - Look at the help for `SummarizedExperiment()` to get more details on the constructor

The code for the answer to question 2A below should contain the following steps:

 1. `Set rownames of pheno_data and convert to class AnnotatedDataFrame(). Convert all factor variables to factors.`
 2. `Set rownames of feature_data and convert to AnnotatedDataFrame()`
 3. `Remove first column of counts, leave only count data`
 4. `Check if rownames(pheno_data) is equal to colnames(counts)`
 5. `Fix rownames and if neccessary`
 6. `create metadata (need to be a list)`
 7. `From the individual components, create a SummarizedExperiment. Add the counts, coldata, metadata and the feature data. You can create the SummarizedExperiment with the function: `SummarizedExperiment()`
 
```{r, results='hide'}

## set rownames of pheno_data, convert to dataframe for rownames 
pheno_data$exposure <- as_factor(as.character(pheno_data$exposure))
pheno_data$cell_type <- as_factor(as.character(pheno_data$cell_type))
pheno_data$condition <- as_factor(as.character(pheno_data$condition))
pheno_data <- as.data.frame(pheno_data)
rownames(pheno_data) <- pheno_data$sample_name
pheno_data

## set rownames of feature_data, convert to dataframe for rownames 
feature_data <- as.data.frame(feature_data)
rownames(feature_data) <- feature_data$ensembl_id
#feature_data <- AnnotatedDataFrame(feature_data)

## colData, convert to dataframe, set rownames, than to matrix
counts <- as.data.frame(counts)
rownames(counts) <- counts$gene_id
counts %>% head(2)
## remove first column to keep only counts
counts <- counts[,-1]
counts %>% head(2)

## convert counts to matrix
counts <- as.matrix(counts)


## check rownames(colData) is equal to colnames(counts)
rownames(pheno_data) == colnames(counts)

## create metadata (needs to be a MIAME formal class object)
#?MIAME
metadata <- list(
  name = c(
    miame$details[1] %>% as.character(), 
    miame$details[2] %>% as.character()),
                 lab = miame$details[3] %>% as.character(),
                 contact = miame$details[4] %>% as.character(),
                 abstract = miame$details[5] %>% as.character(),
                 url = miame$details[6] %>% as.character(),
                 other = miame$details[7] %>% list())

## construct ExpressionSet
#eset <- ExpressionSet(assayData = counts, 
#                      phenoData = AnnotatedDataFrame(pheno_data),
#                      featureData = AnnotatedDataFrame(feature_data),
#                      experimentData = metadata)


# eset

## construct SummarizedExperiment
# se <- makeSummarizedExperimentFromExpressionSet(from = eset)

## create formal class SummarizedExperiment from the indivual parts  
se <- SummarizedExperiment::SummarizedExperiment(
  assays = counts,             ## raw count
  colData = pheno_data,        ## column data
  metadata = metadata,         ## information about exp. lab etc  
  rowData = feature_data)      ## info about the features (genes)
  
```

B) Use the accessors below on your `SummarizedExperiment` object, this is an important check to see if the information is correct and the construction went according plan.

Run the accessors below on your SummarizedExperiment object:

 - `metadata()`
 - `assays()$counts`
 - `colData()`
 
```{r, results='hide'}
metadata(se)
colData(se)
assay(se) %>% head(2)
rowData(se) %>% head(2)
```

# Question 3. Creating an DESeqDataSet 
Createa DESeqDataSet from the SummarizedExperiment object
In order to analyze the data further we need to convert the SummarizedExperiment object to a DESeq2DataSet.
This can be achieved using the `DESeqDataSet()` function.

A) Create an DESeqDataSet from the SummarizedExperiment object below
```{r}
library(DESeq2)
colData(se)

dds <- DESeqDataSet(se, design = ~ condition)
assays(dds)
dds
dds <- DESeq(dds)
```

B) Perform a full differential expression analysis on the DESeqDataSet, create all required plots: p-values, MA, PCA and heatmaps of the top 50 genes, also create an annotated heatmap 
