---
title: "daur2_lesson56_heatmaps"
author: "Marc A.T. Teunis, Alyanne de Haan"
date: '`r Sys.Date()`'
output: 
  html_document: 
    keep_md: yes
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r libs, include=FALSE}

library(tidyverse)
library(SummarizedExperiment)
library(airway)

```

## intro Heatmaps

You are probably already familiar with heatmaps. Most of us have seen them around, as they provide a nice colourfull image of the thousands of datapoints in 'omics datasets.

Heatmaps are graphical representations of data, and each data point (which can of course be an average of more raw data) is placed in a matrix with colour on a colour scale reflecting it's value. So for instance, often more blue-ish colours represent lower values while more red-ish colours represent higher values. This gives a summary of data that is otherwise difficult to summarise. 

In Genomics, heatmaps can be used to visualise the level of expression of many genes across conditions (for instance samples from different patients, cell lines that have or have not been treated with a certain substance etc), all in the same figure. In this case, the colour and intensity of the boxes is used to represent changes  in gene expression (not absolute values). So for instance, upregulated genes get a more red-ish colour.

They will look somewhat like this:

```{r, out.width='50%', fig.align='center', fig.cap='...', echo=FALSE}
knitr::include_graphics(
  here::here(
  'images',
  'heatmap.png')
)
```
(This heatmap was taken from Wikipedia, which generated the heatmap with data extracted from the StemBase database of gene expression data)

When making heatmaps, you can choose how to arrange your data. Often,  each row represents a gene and each column represents a sample. Which gene to put on which row can be based on clustering methods which group genes  together based on the similarity of their gene expression pattern. So for example, all genes that show high expression in samples A,B and C but not D are clustered in the first rows. Then a group of genes that show high expression in samples A and B but not C and D. and so on. This would make it easier to spot general trends in the data.

... In theory. Because in practice reading heatmaps takes some practice.

You have been reading them for sure. Most heatmaps are not as complex as genomic heat maps. But then again, most data is not as complex as genomic data. You will be familiar with these, for instance:

```{r, out.width='50%', fig.align='center', fig.cap='...', echo=FALSE}
knitr::include_graphics(
  here::here(
  'images',
  'covid19_ziekenhuis.png')
)
```

You can easily spot regions that have higher values than others. You can also spot some things to keep in mind while reading and making heatmaps.
Obviously, the spatial position of areas in this heat map will have some correlation with other variables that you may or may not have measured. In the covid-19 hospitalisation map, we do not know if the threshold to be hospitalised may have been different in Jipsingboermussel (county Westerwolde) than in Amsterdam because of the lower population density. Or perhaps the closer personal connection between the "huisarts" and the patients. Also, did they count the residence county, or the hospital county? Does Westerwolde even have a hospital? In this heatmap, areas that are next to each other will even correlate in their value. What do you think, do areas in genomic heatmaps do that as well? Do we always know? Clearly, no. So keep in mind that heatmaps are mainly usefull for giving rough summaries and spotting large trends.

The patterns that are most easily recognised depend a lot on how (on which information) the heatmap designer (that will be you) decided to cluster the data. One additional factor to keep in mind, is that this also depends on the colour scale you use. Large contrasts in colour may over-represent changes in expression, while for instance unicolour heatmaps (changing only colour intensity) may underrepresent them. Also, our brain is really good in comparing contrasts, or colour values that are spatially near each other, but absolutely rubbish at comparing them at a distance. See an excellent example here: https://www.displayr.com/the-problem-with-heatmaps/

Oke, great. So heatmaps are fiddly. Why do we even bother? Because large genomic datasets are quite hard to visualise and summarise with other methods. 

## The basics of genomic heatmaps

Therefore, we will practice a bit with (genomic) heatmaps.
Clustering

We will want to bring the similar rows and columns nearer to each other in the plot. There is no single best strategy to cluster your data, but hierarchical clustering is vary commonly used. It generates the dendrogram you often see on the side of the plot. Basically, this is generated by first finding the pair of samples that is most similar, then finding the next most similar sample, and so on building your way up untill all samples are included.

Similarity here is actually determined by samples having the least distance between them. The distance between two points in a 2-dimensional space (for instance dots on a paper), would be defined by its Euclidean distance, and calculated with a well known formula:

$a^{2} + b^{2} = c^{2}$ 

Now, suppose you have a dataset with 12000 genes and 5 samples, so a data matrix with 12000 rows and 5 columns. Each sample is now defined by a point in 12000-dimensional space, and each gene is defined by a point in a 5 dimensional space. Calculating the Euclidean distance between two points (for example 2 samples) in a many-dimensional space is actually quite similar to doing it in 2 dimensions. And luckily, R: will do it for us with the dist() function.

Suppose we have a dataset with expression data for 5 genes in 4 samples. For the sake of clarity, we will -for now- take a very blunt approach to genetics. Here are the volunteers who supplied us with the samples:

We will make up some count data with some highly expressed genes, some less expressed genes, and one that does not show any counts apart from one sample:

```{r}
gene_hairy1 <- c(200,400,200,400) #more hairy
gene_hairy2 <- c(400,200,400,200) #less hairy
gene_earsize1 <- c(2,20,2,20) # bigger
gene_earsize2 <- c(20,2,20,2) #smaller
gene_wingsize1<- c(0,0,150,0)

expressiondata <- rbind(gene_hairy1,gene_hairy2,
                        gene_earsize1,gene_earsize2,gene_wingsize1)
colnames(expressiondata)<- c("s1_squirtle","s2_eevee","s3_charizard","s4_rattata")
#I would say those things on charizards head are horns
```

Now we can ask R: to calculate a distance matrix for us. It will do this for the rows:

```{r}
dist(expressiondata)

##                gene_hairy1 gene_hairy2 gene_earsize1 gene_earsize2
## gene_hairy2       400.0000                                        
## gene_earsize1     605.9769    617.7443                            
## gene_earsize2     617.7443    605.9769       36.0000              
## gene_wingsize1    602.0797    550.0000      150.6917      131.5599
```

The least distant two genes are B1 and B2, so those will be clustered together. We can plot a dendrogram of the distance matrix:

```{r}
dist(expressiondata) %>% hclust() %>% plot()
```


We can see that the expression of gene hairy1 and hairy2 are more similar to each other than to the other genes, earsize1 and earsize2 are similar, and the expression of wingsize1 looks a bit more like that of the earsize genes than to that of the hairy genes. Looking at the data in expressiondata, that seems reasonable at first glance.

Now let’s make a heatmap. We will use the pheatmap package:

```{r}
#install.packages("pheatmap")
library(pheatmap)

pheatmap(expressiondata)
```


heatmap1-1.png

On top, you see the function’s clustering of the samples. You can see the difference in hairy genes expression pattern between samples 1 and 4 versus samples 2 and 3.

The dendogram on the left of the figure shows the clustering we just calculated as well with dist(). You can see that the lowly expressed genes have somewhat blue-ish rows, and the highly expressed hairy genes show more contrasts. We can not actually make out much more than this, the top 3 rows are not informative at all.

 
Scaling

We are not in fact interested in which genes have high expression and which have low expression in general, but rather in differences in expression of the 5 genes over the samples.

The clustering of genes in this heatmap also seems to depend rather a lot on general expression levels. This is because the clustering method we used (Euclidean distance) just takes absolute difference between data points. Therefore, the hairy genes are clustered, so are the earsize genes, and wingsize1 is clustered with the earsize genes. Looking at the pokemon who provided the samples, you can see that this might not be the most informative clustering. When doing sequencing analysis, we often use different clustering methods. There are many, for now it is enough to know the basics.

Usually, genomic heatmaps are scaled by rows. This means that we recalculate the data in the expression matrix to reflect expression of a gene relative to its expression in other samples (columns) rather than absolute values. Many heatmap functions do this automatically, so take care if you use another function (there are many heatmap functions, we just use pheatmap() because we like this one). pheatmap() does not scale by default.

We can ask it to center and scale the values though, it does so by subtracting the mean (centering) and dividing by the standard deviation (scaling), thus calculating Z scores:

```{r}
pheatmap(expressiondata, scale="row")
# heatmapscaled-1.png

```


Now the rows are easier to compare, and we see that the expression of hairy1 and earsize1 seems to be correlated, as well as the hairy2 and earsize2. Indeed, the more hairy pokemon in our selection tend to have smaller ears and the less hairy pokemon tend to have bigger ears. Also, the expression of the wingsize gene seems to be correlated to expression of hairy1 and earsize1 genes.

This plot also illustrates nicely that heatmap clusters are based on the data, not on whether we know that the genes hairy1 and hairy2 have something to do with each other.

Now, we took some pretty nice fictional count data. In reality, the counts will differ a bit.

```{r}
# add some noise for illustrative purposes

set.seed(1)


noise <- matrix(runif(5*4, min=0.8, max=1.2), 5, 4)
expressiondata2 <- round(expressiondata*noise)
pheatmap(expressiondata2, scale="row")

#heatmapnoised-1.png
```
 

That is starting to look like a proper heatmap.

You can change the colours used in the heatmap. Try to play around with the following suggestions. What does the number do (e.g. rainbow(30) vs rainbow(3))?

```{r}
pheatmap(expressiondata2, scale="row",color=rainbow(30))
pheatmap(expressiondata2, scale="row",color=rainbow(3))
pheatmap(expressiondata2, scale="row", color = c("black", "red"),
         breaks = c(-1.5, 0, 1.5),)

# or setting the colours beforehand, in this case with the popular, unreadable green-black-red gradient:
cols = colorRampPalette(c("green", "black", "red"))(30)
pheatmap(expressiondata2, scale="row",color=cols)

# or even use colour palette packages (which you need to install first)
library(viridis)
pheatmap(expressiondata2, scale="row",color=inferno(30))
```
 
Reading heat maps

Let’s look at some heat maps from papers. I googled “heat map gene expression” and found this one as one of the top images:

micelizards.png

image from: Sarkar H, Arya S, Rai U, Majumdar SS(2016) A Study of Differential Expression of TesticularGenes in Various Reproductive Phases ofHemidactylus flaviviridis (Wall Lizard) to Derive TheirAssociation with Onset of Spermatogenesis and ItsRelevance to Mammals. PLoS ONE 11(3): e0151150. doi:10.1371/journal.pone.0151150 .

It shows a differential gene expression pattern in pre-meiotic, meiotic and post-meiotic testes of wall lizard and mice, and you can see that for these genes, there seems to be a buildup, with increased expression mostly in the active reproductive phase of the Indian wall lizard (wall lizards have breeding seasons).

 
Next up in the series of random heat maps we grabbed from open access papers for you to practice with: tomatoes.
tomato.png

Zouine M, Fu Y, Chateigner-Boutin A-L, Mila I, Frasse P, et al. (2014) Characterization of the Tomato ARF Gene Family Uncovers a Multi-Levels Post-Transcriptional Regulation Including Alternative Splicing. PLoS ONE 9(1): e84203. doi:10.1371/journal.pone.008420

Hopefully, you can see that the authors of this study found that group I genes are expressed in roots while group II genes are less so. Several of the genes in both groups roughly seem to lower in expression when a tomato starts to ripen (last 4 columns), so perhaps these have some role in fruit development.

Now take a look at this one, taken from a study on diabetes, with yellow for upregulated and red for downregulated gene expression (why?! don’t do this) in two groups of pregnant participants: a group that screened positive for gestational diabetes but turned out to be non-diabetic (ND) and a group who screened negative (and did not have diabetes either) (C):
diabeticheatmap.png

Gelaleti RB,Damasceno DC,Salvadori DMF, et al. Geneexpression profile of wholeblood cells differs inpregnant women withpositive screening andnegative diagnosis forgestational diabetes. BMJOpen Diabetes Research andCare 2016;4:e000273.doi:10.1136/bmjdrc-2016-000273 .

By now, you should be able to tell why the researchers did not just plot all the people in the ND-group next to each other.

Over the next page, you will learn how to generate your own genomic heatmaps, and analyse RNA sequencing data using  DESeq2.
