---
title: "Analyzing RNA-seq data with DESeq2"
author: "Marc A.T. Teunis (Adapted after Michael I. Love, Simon Anders, and Wolfgang Huber)"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: >
  A basic task in the analysis of count data from RNA-seq is the
  detection of differentially expressed genes. The count data are
  presented as a table which reports, for each sample, the number of
  sequence fragments that have been assigned to each gene. Analogous
  data also arise for other assay types, including comparative ChIP-Seq,
  HiC, shRNA screening, mass spectrometry.  An important analysis
  question is the quantification and statistical inference of systematic
  changes between conditions, as compared to within-condition
  variability. The package DESeq2 provides methods to test for
  differential expression by use of negative binomial generalized linear
  models; the estimates of dispersion and logarithmic fold changes
  incorporate data-driven prior distributions This vignette explains the
  use of the package and demonstrates typical workflows.
  [An RNA-seq workflow](http://www.bioconductor.org/help/workflows/rnaseqGene/)
  on the Bioconductor website covers similar material to this vignette
  but at a slower pace, including the generation of count matrices from
  FASTQ files.
  DESeq2 package version: `r packageVersion("DESeq2")`
output: html_document
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
```	

# Standard workflow

**Note:** if you use DESeq2 in published research, please cite:

> Love, M.I., Huber, W., Anders, S. (2014)
> Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.
> *Genome Biology*, **15**:550.
> [10.1186/s13059-014-0550-8](http://dx.doi.org/10.1186/s13059-014-0550-8)

Other Bioconductor packages with similar aims are
[edgeR](http://bioconductor.org/packages/edgeR),
[limma](http://bioconductor.org/packages/limma),
[DSS](http://bioconductor.org/packages/DSS),
[EBSeq](http://bioconductor.org/packages/EBSeq), and 
[baySeq](http://bioconductor.org/packages/baySeq).

## Packages
```{r}
library(SummarizedExperiment)
library(DESeq2)
# BiocManager::install("pasilla")
library(pasilla)
library(GEOquery)
library(tidyverse)
```

## Quick start
Here we show the most basic steps for a differential expression
analysis. There are a variety of steps upstream of DESeq2 that result
in the generation of counts or estimated counts for each sample, which
we will discuss in the sections below. This code chunk assumes that
you have a count matrix called `cts` and a table of sample
information called `coldata`.  The `design` indicates how to model the
samples, here, that we want to measure the effect of the condition,
controlling for batch differences. The two factor variables `batch`
and `condition` should  be columns of `coldata`. 
The most easy startpoint to perform this analysis is if you have a SummarizedExperiment class object that stores all the neccessary data.

```{r quickStart, eval=FALSE}
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design= ~ batch + condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
res <- results(dds, name="condition_trt_vs_untrt")
# or to shrink log fold changes association with condition:
res <- lfcShrink(dds, coef="condition_trt_vs_untrt", type="apeglm")
```

The following assumptions are made:

 - We focus on transcriptomics analysis only in this lessons
 - We assume you have the raw RNAseq counts available in a matrix, where the columns are the sample ids and the rows are the genes from which the transcripts

## How to get help for DESeq2

Any and all DESeq2 questions should be posted to the 
**Bioconductor support site**, which serves as a searchable knowledge
base of questions and answers:

<https://support.bioconductor.org>

## Input data

### Why un-normalized counts?

As input, the DESeq2 package expects count data as obtained, e.g.,
from RNA-seq or another high-throughput sequencing experiment, in the form of a matrix of integer values. The value in the *i*-th row and the *j*-th column of the matrix tells how many reads can be assigned to gene *i* in sample *j*. Analogously, for other types of assays, the rows of the matrix might correspond e.g. to binding regions (with ChIP-Seq) or peptide sequences (with quantitative mass spectrometry). We will list method for obtaining count matrices in sections below.

The values in the matrix should be un-normalized counts or estimated
counts of sequencing reads (for
single-end RNA-seq) or fragments (for paired-end RNA-seq). 
The [RNA-seq workflow](http://www.bioconductor.org/help/workflows/rnaseqGene/) describes multiple techniques for preparing such count matrices.  It
is important to provide count matrices as input for DESeq2's statistical model [@Love2014] to hold, as only the count values allow assessing the measurement precision correctly. The DESeq2 model internally corrects for library size, so transformed or normalized values such as counts scaled by library size should not be used as
input.

### The DESeqDataSet

The object class used by the DESeq2 package to store the read counts 
and the intermediate estimated quantities during statistical analysis
is the *DESeqDataSet*, which will usually be represented in the code
here as an object `dds`.

A technical detail is that the *DESeqDataSet* class extends the
*(Ranged)SummarizedExperiment* class of the 
[SummarizedExperiment](http://bioconductor.org/packages/SummarizedExperiment) package. The "Ranged" part refers to the fact that the rows of the assay data (here, the counts) can be associated with genomic ranges (the exons of genes). This association facilitates downstream exploration of results, making use of other Bioconductor packages' range-based functionality
(e.g. find the closest ChIP-seq peaks to the differentially expressed genes).

A *DESeqDataSet* object must have an associated *design formula*.
The design formula expresses the variables which will be
used in modeling. The formula should be a tilde (~) followed by the
variables with plus signs between them (it will be coerced into an
*formula* if it is not already). The design can be changed later, 
however then all differential analysis steps should be repeated, 
as the design formula is used to estimate the dispersions and 
to estimate the log2 fold changes of the model. 

*Note*: In order to benefit from the default settings of the
package, you should put the variable of interest at the end of the
formula and make sure the control level is the first level.

We will now show constructing a *DESeqDataSet*, depending
on what pipeline was used upstream of DESeq2 to generated counts or
estimated counts you could use another: see the DESeq2 vignette or RNAseq workflow on bioconductor for more details.

We will start from downloading a dataset from GEO with the GEOqeary packages (See lesson 4)

## Example data
In this demo we will look at RNAsequencing data from an experiment with Zebrafish

RNA-Seq data processing and analysis.

Quote from the article by Weger et al., 2016

Image analysis and base calling were performed using the Illumina pipeline (Real Time Analysis and Offline Base Caller 1.8). All raw sequence data are available from the GEO database (accession number GSE76073, http://www.ncbi.nlm.nih.gov/geo/). Reads were mapped onto the Zv9 assembly of the zebrafish genome using STAR 2.3.8 [35]. Uniquely mapped reads were counted for each gene locus as annotated in Ensembl v75 using the Python Package HTseq v0.5.3p1. Read counts were normalized using DESeq2 [36]. For downstream analysis a variance stabilizing transformation was applied to the normalized data [37].

To complete this exercise you need to perform a differential expression analysis on the raw count data from the Weger et al., 2016 paper. The demo that we have seen on RNAseq data anlysis can be used as a guide for the steps you will have to complete. A summary of the crude steps is given below. The final end-product is a heatmap of the top 20 differetially expressed genes for the different treatments in the paper.
Publication and Data

https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006512#sec010

Be aware: the experiment consists of two seperate RNAseq experiments (“GPL15583” and “GPL16509”)

Here we only analyze the first experiment “GPL15583” and we focus on the wild type zebrafish only, for simplicity.

### Download dataset from GEO
```{r}
gds_zebrafish <- getGEO("GSE76073")
## inspect 
gds_zebrafish # a list with 1 element
## extract the Expression set for later use
eset <- gds_zebrafish[[1]]

names(gds_zebrafish)

exprs(eset)

```

### Getting the raw counts 
The downloaded ExpressionSet from GEO does not include the raw counts that we need to analyze this data. For this dataset the raw counts are available as asupplementary file. This takes a while
```{r}
getGEOSuppFiles("GSE76073")

```

### Load the raw counts into R
```{r}

cnts <- read.delim(
  here::here(
    "GSE76073",
    "GSE76073_Raw_counts.txt.gz"
  ), stringsAsFactors = FALSE
) 

head(cnts) %>% as_tibble()

```

### Getting phenotypic (sample) data
This is the information on the samples that relate to the columns of the cnts dataframe. You need to look carefully at this table to learn the exp design and to fix the colnames tof the cnts dataframe.
```{r}
col_data <- pData(eset)
rownames(col_data) ## these are the series numbers

```

### colnames of the cnts dataframe
The colnames of the cnts matrix are a combination of genotype, treatment and zeitgeber. In order to match the right sample_names (rowname) of colData with the right column name of `cnts` I create a new col_data dataframe with less columns and an addiotional column that is a united column of genetype, treatment and zeitgeber. I will use lowercase and set the colnames of the `cnts` also to lowercase. Than we add the new united column (after some cleaning and renaming) as rownames to the new coldata object. We select the proper columns in `cnts` and matches the order of the columns to the rownames of col_data_new. 

```{r}
colnames(cnts)
rownames(col_data)
```

**Conclusion, we need to fix the colnames of `cnts` in order for it to match the rownames of `col_data`

### Construct a SummarizedExperiment
The essential step here is that the columnames of the counts DataFrame should exactly match (in order and symantics) the rownames of the colData element.

If all the elements are in the right shape we can create a SummarizedExperiment with `SummarizedExperiment()`

To remind you of the relationships between the different elements of a formal class `SummarizedExperiment` R object:
```{r}

```

These are the steps:
 1. Set rownames of pheno_data and convert to class `AnnotatedDataFrame()`. Convert all factor variables to factors.`
 1. `Set rownames of feature_data and convert to AnnotatedDataFrame()`
 1. `Remove first column of counts, leave only count data`
 1. `Check if rownames(pheno_data) is equal to colnames(counts)`
 1. `Fix rownames and if neccessary`
 1. `create metadata (need to be a list)`
 1. Add `featureData`
 1. Check the factor levels of the experimental predictors/grouping variables, and relevel if neccessary
 1. From the individual parts create a `SummarizedExperiment`. Add the metadata. You can create the SummarizedExperiment from the ExpressionSet with the function:
`SummarizedExperiment()`

look at the help function for `SummarizedExperiment()` to learn what arguments should be passed in

```
?SummarizedExperiment
```

To complete these steps we will need tools from bioconductor packages, for example to access data stored in the `ExpressionSet` that we downloaded from GEO. The data wrangling that is needed to harmonize the colnames and the rownames can be done with tools from `{tidyverse}`.

### colData
First we extract the colData (or so called phenoData in an `ExpressionSet`) from the expression set. This object contains information about the experimental samples, so it is the basis for the colData object in our new `SummarizedExperiment` object.
```{r}
col_data <- pData(eset)
col_data %>% as_tibble()
```

**Conclusion** In this particular phenoData object we see information about the samples, the experimental design but also some meta-data like the publshing data, the contact email-address, the contact-address etc. This meta data can be used to fill the slot for experimentData in our new 

### Clean and filter colData to match column names of `cnts`
We also clean the names of this dataframe with `{janitor}`
```{r}
names(col_data) <- names(col_data) %>% janitor::make_clean_names()

col_data %>%
  dplyr::select(
    title,
    geo_accession,
    genotype_ch1,
    strain_ch1,
    treatment_ch1, 
    zeitgeber_time_ch1
  
) -> col_data_new

rownames(col_data_new)
col_data_new %>% as_tibble()
```

### Match colnames of `cnts` with rownames of `col_data_new`
There seems to be one column in `col_data_new` that sort-of resembles the colnames in `cnts`. Only it does not contain info on the treatment. 
```{r}
colnames(cnts)
col_data_new$title %>% as.character()
```

### It easier to create a new column in col_data_new on the basis of columns `genotype_ch1`, `treatment_ch1` and `zeitgeber_time_ch1`. We have to first change the values, for these variables for it to match the keys in the colnames of `cnts`  
```{r}

names(col_data_new)
col_data_new <- col_data_new %>%
  dplyr::mutate(
    genotype_ch1 = ifelse(
      genotype_ch1 == "Wild type", "WT", genotype_ch1),
    genotype_ch1 = ifelse(
      genotype_ch1 == "rx3 weak mutant", "WEAK", genotype_ch1),
    genotype_ch1 = ifelse(
      genotype_ch1 == "rx3 strong mutant", "RX3", genotype_ch1),
    ## we managed to come close, no change treatment
    treatment_ch1 = "DMSO" ## only a single value we need to change
    ## zeitgeber is already the good key

  ) %>%  
## now we create a new column
  unite(
    col = "sample_id",
    c("genotype_ch1",
      "treatment_ch1",
      "zeitgeber_time_ch1"),
    remove = FALSE,
    sep = "_"
    )

col_data_new$sample_id
colnames(cnts)
```

### Match colnames of `cnts` with rownames of `col_data_new`

Steps:

 - First we select the proper columns from cnts that belong to the coldata, 
 - We change the order of cnts columns to match the rows in `col_data_new`
 - than we check it all worked out
 - Finally we are ready to create our SummarizedExperiment, add some metadata just for kicks, and we can continue.

**You see, data cleaning and preparation indeed takes about 70-80% of your work. If only the people who are submitting the data knew more about the underlying principles!! 

```{r}
## set new rownames on col_data_new
rownames(col_data_new) <- col_data_new$sample_id
## check manually if this went well

## select columns of cnts that we have colData for
ind_col <- colnames(cnts) %in% rownames(col_data_new) ## should be length(cnts)

## we can remove column 21 'the gene length'
## let's put it in a featureData dataframe, together with the rownames of cnts as rownames for featureData
feature_data <- data.frame(
  gene_length = cnts$gene_length
)
rownames(feature_data) <- rownames(cnts)

## select the proper columns in cnts
cnts_new <- cnts[, ind_col]

```

### Are we done?
To see if we are ready to create our `SummarizedExperiment` we check if the rownames of col_data_new match the colnames of `cnts_new`
```{r}
colnames(cnts_new) == rownames(col_data_new)

```

**YES!! WE ARE DONE** We are lucky that we do not have to change the column order of `cnts_new`
If you are not so lucky: here is pointer on how to do this:

```{r}
## let's reorder the order of col_data_new
col_data_dummy <- col_data_new[c(12:1), ]

col_data_dummy$sample_id == col_data_new$sample_id ## the order is reverse
colnames(cnts_new) == rownames(col_data_dummy)

## using the new order to rearrange the columns in cnts_new
cnts_dummy <- cnts_new %>%
  dplyr::select(
    col_data_dummy$sample_id
  )

colnames(cnts_dummy) == rownames(col_data_dummy) ## we used the values in `sample_id` to 'select' the columns of cnts, using dplyr::select()


```

### Final check for colnames `cnts` vs `rownames col_data_new`
```{r}
## check rownames(colData) is equal to colnames(counts)
rownames(col_data_new) == colnames(cnts_new)

```

### Add metadata
```{r}


## create metadata (needs to be a list)
## we use the  original col_data table to get a bit of metadata
metadata <- list(
  name = col_data$contact_name %>% unique(),
  email = col_data$contact_email %>% unique(),
  accessions = col_data$geo_accession %>% unique()
)
```

### Constructing the `SummarizedExperiment` object
```{r}
# proteomic
# metabolomics
# microbiome analysis
# genomic sequencing
# transcriptomics
# kinomics

sumexp <- SummarizedExperiment(
  assays = as.matrix(cnts_new), # supplementary data downloaded
  colData = col_data_new, # part of the expression set via GEOquery
  rowData = feature_data, # rownames van de counts matrix, addiotional column
  metadata = metadata # part of the colData
) 

sumexp

## access the counts
head(assay(sumexp))
colData(sumexp)

readr::write_rds(sumexp, path = here::here("data", "sumExp_lesson6-2.rds")) 
readr::read_rds(path = here::here("data", "sumExp_lesson6-2.rds"))

assay(sumexp) %>% as_tibble()

```

## Differential expression analysis
We use the DESeq2 package 

### Prepare analysis using the `SummarizedExperiment` object and the `DESeqDataSet()` function. We look at the effect of `genotype_ch1` and `zeitgeber_time_ch1` on the differential gene expression by passing in the formula

`design = ~ genotype_ch1 + zeitgeber_time_ch1`

```{r}
## Check the levels of the exeperimental factors before moving on
## factor relevellen: see chapter 'forcats / Forcats' uit R4DS!!
col_data_new$genotype_ch1 %>% as.factor() %>% levels
col_data_new$zeitgeber_time_ch1 %>% as.factor() %>% levels

ddsSE <- DESeq2::DESeqDataSet(sumexp, design = ~ genotype_ch1 + zeitgeber_time_ch1)

```

### We can filter for low expression before analyzing further
Only genes with total row count sum of more than 10 are kept in the data.
```{r}
keep <- rowSums(counts(ddsSE)) >= 24 ## filter for low abundance
ddsSE <- ddsSE[keep,]

ddsSE
```

### Running the analysis with default settings
```{r}
ddsDE <- DESeq2::DESeq(ddsSE) ## normalisation and scaling is included
```

### Getting the result table
The results table is the table containing the end results of the analysis. This is what you should write to disk for future reference, along with the code that generated the table. Store them close together. Preferably you build an R-package around your analysis, containing all the components neccessary to reproduce the exact result. In the module 'Workflow' that will be part of the bioinformatics courses from 2021 onwards, we will demonstrate how to build such an R package.
```{r}
res <- results(ddsDE) 

res_df <- res %>% as_tibble()

## top genes
res[order(res$padj), ] %>% head

names(res)

## getting results for more genotype comparisons
res_wt_vs_weak <- results(ddsDE, contrast=c("genotype_ch1","WT","WEAK"))
res_wt_vs_rx3 <- results(ddsDE, contrast=c("genotype_ch1","WT","RX3"))
res_zt21_vs_zt03 <- results(ddsDE, contrast=c("zeitgeber_time_ch1","ZT21","ZT03"))
```
From the result table we can see two comparisons:

 1. The `WT` condition, for all zeitgeber_ch1 conditions vs the weak mutant condition `WEAK`, for all zeitgeber_ch1 conditions. 
 1. The `WT` condition, for all zeitgeber_ch1 conditions vs the weak mutant condition `RX3`, for all zeitgeber_ch1 conditions.
 
Because there is only one condition per sample, and no replicates, we have to be careful on drawing strong statistical conclusions. 

## Log fold change shrinkage for visualization and ranking

Shrinkage of effect size (LFC estimates) is useful for visualization and ranking of genes. To shrink the LFC, we pass the dds object to the function lfcShrink. Below we specify to use the apeglm method for effect size shrinkage (Zhu, Ibrahim, and Love 2018), which improves on the previous estimator.

We provide the dds object and the name or number of the coefficient we want to shrink, where the number refers to the order of the coefficient as it appears in resultsNames(dds).
```{r}
resultsNames(ddsDE)
## [1] "Intercept"                      "condition_treated_vs_untreated"

resLFC <- lfcShrink(ddsDE, coef="genotype_ch1_WT_vs_RX3", type="apeglm")
resLFC
```

## Result summary
```{r}
summary(res)
```

## Plotting the results

The are at least 4 plot-type you should generate on the results:

 1. Distribution of p-values
 1. The MA plot showing and MA plot with reduced noice (shrunken log2 fold-changes )
 1. PCA plot
 1. Heatmaps
 
### Look at the p-values
Here we look at the
```{r}
res %>%
  as_tibble()

ggplot(as(res, "data.frame"), aes(x = pvalue)) +
  geom_histogram()
```

### MA Plots
In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored red if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.
```{r}
DESeq2::plotMA(ddsDE, ylim = c(-2, 2))
```

It is more useful visualize the MA-plot for the shrunken log2 fold changes, which remove the noise associated with log2 fold changes from low count genes without requiring arbitrary filtering thresholds.
```{r}
plotMA(resLFC, ylim=c(-2,2))
```

## Principal Component Analysis
Provides information on how the samples are different
```{r}
ddsDE_rlog <- ddsDE %>%
  rlogTransformation() 

ddsDE_rlog %>%
  DESeq2::plotPCA(intgroup = c("genotype_ch1", "zeitgeber_time_ch1")) +
  geom_vline(xintercept = 0, linetype = "dashed", colour = "red") +
  geom_hline(yintercept = 0, linetype = "dashed", colour = "red")

```
**Conclusion** There is a clear effect of genotype and zeitgeber on the differences in gene-expression, considering the point are distributed away from the centerpoint where PC1 = 0 and PC2 = 0 (crosshair dotted red lines) 

### Heatmaps
To learn more on the gene expression differences, a heatmap can be very effective to communicate the results. 
```{r, fig.height=12, fig.width=10}
library(pheatmap)
matrix_results <- assay(ddsDE_rlog)

ind <- matrix_results %>%
  rowMeans() %>%
  order(decreasing = TRUE) ## only the first 30 genes

top30 <- matrix_results[ind[1:30],]

heatmap <- pheatmap(top30,
                    scale = "row",
                    annotation_col = as.data.frame(
                      colData(ddsDE_rlog)[, c("genotype_ch1",
                                              "zeitgeber_time_ch1")]
                    ))
heatmap
```

### Annotated heatmap
As a final step we will create an annotated version of the heatmap. Above we have the heatmap displaying the top 30 differentially regeulated genes with their respective ENSEMBL Zebrafish ids. To better understand the biological significance we would prefer to have gene names in stead. 
There are many way to get annotations for these genes. We alreday know one way: using a `{biomaRt}` query.

We will have to use the specific Zebrafish dataset to get the right annotations for the zebrafish ENSEMBL ids. The Latin name for zebrafish is _Danio rerio_ or _D. rerio_ in short. 

```{r}
library(biomaRt)
## all available marts
listMarts()

## using a mart
mart <- biomaRt::useMart('ensembl')

## get all available datasets
datasets <- listDatasets(mart) %>% 
  as_tibble()

## index for matching row that contain "Zebrafish" or "zebrafish"
ind <- str_detect(string = datasets$description, pattern = "(Z|z)ebrafish")

## get records
danio <- datasets[ind,]

## use the right mart
mart <- biomaRt::useMart('ensembl', dataset = danio$dataset)

attributes <- listAttributes(mart = mart) %>% as_tibble()

## search for gene_name or symbol
ind_attrib <- str_detect(
  string = attributes$name,
  pattern = "(gene_name|symbol)"
)

attributes[ind_attrib,]

annotations <- 
  getBM(
    attributes=c('zfin_id_symbol', 'ensembl_gene_id'),
    filters = 'ensembl_gene_id', 
    values = rownames(top30), 
    mart = mart)

```

**Conclusion** On the basis of the top30 genes and their respective ensembl_ids, we find 27 annotated genes. Let's plot a new heatmap showing those with the 27 new names 

```{r}
top30_df <- top30 %>%
  as.data.frame() %>%
  dplyr::mutate(ensembl_gene_id = rownames(.))
  
annot_27 <- left_join(annotations, top30_df)

top_27_annot <- as.matrix(annot_27[, -c(1:2)])
rownames(top_27_annot) <- annot_27$zfin_id_symbol

```

## Annotated heatmap
```{r, fig.height=12, fig.width=10}
heatmap <- pheatmap(top_27_annot, 
                    scale = "row",
                    annotation_col = as.data.frame(
                      colData(ddsDE_rlog)[, c("genotype_ch1",
                                              "zeitgeber_time_ch1")]))
                    
heatmap
```

## Next steps:

 - pathway analysis
